### ISSUE TITLE
Left over vEthernet adapters and persistent routes

### ISSUE THREAD

**Issue by mihail-velikov** [openvswitch/ovn-kubernetes#360]
Hello,

Our windows nodes recently started to lose network connectivity shortly after a clean restart and starting up the kubelet. That seemed very strange behavior and for a little while I continued to only restart the node and the service for docker and kubelet. Then I started to notice that the networking on the machine started to take a couple of minutes to start after the machine boots up. After some investigation I found out that I had probably around a hundred persistent routes that looked like this:

===========================================================================
Persistent Routes:
  Network Address          Netmask  Gateway Address  Metric
          0.0.0.0          0.0.0.0      10.31.0.254       1
        10.32.0.0      255.255.0.0        10.32.2.1       2
      10.32.200.0    255.255.255.0        10.32.2.1       2
          0.0.0.0          0.0.0.0      10.31.0.254  Default
          0.0.0.0          0.0.0.0        10.32.2.1  Default
          0.0.0.0          0.0.0.0        10.32.2.1  Default
          0.0.0.0          0.0.0.0        10.32.2.1  Default
          0.0.0.0          0.0.0.0        10.32.2.1  Default
          0.0.0.0          0.0.0.0        10.32.2.1  Default

This seemed very interesting as after a reboot I start docker, clean all containers and then start the kubelet:
1. PS C:\Program Files\Docker> .\dockerd.exe --config-file C:\ProgramData\docker\config\daemon.json -D > daemon.log 2>&1
2. PS C:\docker rm -f $(docker ps -a -q)
3. start kubelet in another powershell

Another thing that I noticed was there were multiple hyperv vEthernet adapters even with no containers running:
![vmnetworkadapters](https://user-images.githubusercontent.com/13517115/41097938-72766c3a-6a62-11e8-9e8e-30b67cf1bd8a.png)

![ipconfig](https://user-images.githubusercontent.com/13517115/41098259-76ab63b8-6a63-11e8-8598-a12e621b3fe0.png)

I also noticed that when containers are added more of the persistent routes are also added and my guess is that when force deleting containers after boot the routes are not cleaned. At some point this leads to a huge amount of routes lingering which then breaks the networking on the machine.

Any input would be greatly appreciated.

Btw - after deleting the routes the networking is stable again.

Forgot to mention details about my environment:
windows version:
PS C:\Windows\system32> docker --version
Docker version master-dockerproject-2018-05-20, build 824ce49a

OVN: 2.7.0

Kubernetes version: 1.10.0

Thank you.

Kind Regards,
Mihail Velikov

**Comment by aserdean** [comment]
Creating a vEthernet adapter also adds persistent routes to 0.0.0.0 for that interface, starting from 1709 (no idea why :) ), I posted a PR: https://github.com/openvswitch/ovn-kubernetes/pull/356 which should take care of it on the non Hyper-V scenario.

What version of Windows are you using?
Alin.

**Comment by mihail-velikov** [comment]
Hello Alin,

I guess I forgot to paste the windows version :) . We are using Windows 2019 Preview:
C:\Windows\system32>ver
Microsoft Windows [Version 10.0.17623.1002]

Another interesting thing is that we have an adapter called (it is visible in the screenshot):
Hyper-V Virtual Ethernet Adapter #96 

Not sure if this is relevant :)

Kind Regards,
Mihail Velikov

**Comment by aserdean** [comment]
Hmmz.

Can you provide the output of the following:
`Get-NetAdapter -IncludeHidden`
`Get-HnsEndpoint`

**Comment by mihail-velikov** [comment]
![netadapter1](https://user-images.githubusercontent.com/13517115/41102750-417ccbc0-6a70-11e8-86dc-28f78df6d369.png)
![netadapter2](https://user-images.githubusercontent.com/13517115/41102751-436404a8-6a70-11e8-9a18-b1207cdebf4c.png)
![endpoint1](https://user-images.githubusercontent.com/13517115/41102753-45608d12-6a70-11e8-84de-65766d8faca7.png)
![endpoint2](https://user-images.githubusercontent.com/13517115/41102756-47ba2230-6a70-11e8-9318-81473a2abec5.png)
![endpoint3](https://user-images.githubusercontent.com/13517115/41102771-4d1ff0c4-6a70-11e8-94be-8e5ff1925299.png)
![endpoint4](https://user-images.githubusercontent.com/13517115/41102780-4fa3e0b2-6a70-11e8-9320-31c06620e5df.png)

This is one of the nodes. The other nodes is reported by kubernetes as notReady but containers are still running. There I can get the list of the adapters but Get-HnsEndpoint does not return anything

**Comment by aserdean** [comment]
From the look of it you have a lot of dead adapters. Are they still persistent across reboots?
Also can you try to remove them by using `Remove-VMNetworkAdapter` if you have the Hyper-V feature installed?
I think that you can't even delete them, but just want a confirmation.

**Comment by mihail-velikov** [comment]
Hello Alin,

Yes the adapters are persistent across reboots and I did try to delete them. Unfortunately it was not possible. I believe it was an "access denied" error but I don't have a screenshot. Since then I had to reinstall the machines in order to restore the functionality. I am currently redeploying one of our debug cluster and I will update to the latest version of ovnkube. Afterwards I will try to replicate the behavior.

Kind Regards,
Mihail Velikov
